cmd/internal/kube/daemonset/daemonset.go

```diff
@@ -135,48 +227,143 @@ func (d *Daemonset) KubernetesEventsWorker() {
+		if podStatus == string(coreV1.PodSucceeded) || podStatus == string(coreV1.PodFailed) {
+			printer.Infof("pod %s has stopped running", podStatus)
+
+			podArgs, err := d.getPodArgsFromMap(podName)
+			if err != nil {
+				printer.Errorf("failed to get podArgs for pod %s: %v", podName, err)
+				continue
+			}
+
+			isSameStage, err := podArgs.validatePodTrafficMonitorStage(PodTerminated, TrafficMonitoringStarted)
+			if err != nil {
+				printer.Errorf("pod %s is in invalid state %d", podName, podArgs.PodTrafficMonitorStage)
+			}
+			if isSameStage {
+				printer.Errorf("pod %s is already in state %d", podName, podArgs.PodTrafficMonitorStage)
+			}
+
+			podArgs.setPodTrafficMonitorStage(PodTerminated)
+
+			d.StopApiDumpProcess(
+				podName,
+				fmt.Errorf("pod %s has stopped running, status: %s", podName, podStatus),
+			)
+		}
+	}
+}
+
+func (d *Daemonset) PodsHealthWorker(done <-chan struct{}) {
+	if d.PodHealthCheckInterval <= 0 {
+		return
+	}
+
+	ticker := time.NewTicker(d.PodHealthCheckInterval)
+	for {
+		select {
+		case <-done:
+			return
+		case <-ticker.C:
+			d.checkPodsHealth()
+		}
+	}
+
 }
 
-func (d *Daemonset) StartApiDumpProcess(podArgs PodArgs) error {
+func (d *Daemonset) StartApiDumpProcess(podName string) error {
+	podArgs, err := d.getPodArgsFromMap(podName)
+	if err != nil {
+		return err
+	}
+
+	isSameStage, err := podArgs.validatePodTrafficMonitorStage(TrafficMonitoringStarted, PodDetected, PodInitialized)
+	if isSameStage || err != nil {
+		return err
+	}
+
 	networkNamespace, err := d.CRIClient.GetNetworkNamespace(podArgs.ContainerUUID)
 	if err != nil {
-		return errors.Wrapf(err, "failed to get network namespace for pod/containerUUID: %s/%s", podArgs.PodName, podArgs.ContainerUUID)
+		return errors.Wrapf(err, "failed to get network namespace for pod/containerUUID: %s/%s",
+			podArgs.PodName, podArgs.ContainerUUID)
 	}
 
 	// Channel to stop the API dump process
 	stopChan := make(chan error)
 
 	apidumpArgs := apidump.Args{
-		ClientID:                  telemetry.GetClientID(),
-		Domain:                    rest.Domain,
-		ServiceID:                 podArgs.InsightsProjectID,
-		TargetNetworkNamespaceOpt: optionals.Some(networkNamespace),
-		ReproMode:                 podArgs.InsightsReproModeEnabled,
-		StopChan:                  stopChan,
+		ClientID:  telemetry.GetClientID(),
+		Domain:    rest.Domain,
+		ServiceID: podArgs.InsightsProjectID,
+		ReproMode: d.InsightsReproModeEnabled,
+		DaemonsetArgs: optionals.Some(apidump.DaemonsetArgs{
+			TargetNetworkNamespaceOpt: networkNamespace,
+			StopChan:                  stopChan,
+			APIKey:                    podArgs.PodCreds.InsightsAPIKey,
+			Environment:               podArgs.PodCreds.InsightsEnvironment,
+		}),
 	}
 
-	// Put the process stop channel map and start the process in separate go routine
-	d.PodNameStopChanMap.Put(podArgs.PodName, stopChan)
+	podArgs.StopChan = stopChan
+	podArgs.setPodTrafficMonitorStage(TrafficMonitoringStarted)
```

Comment:
No mutex held here.  The model to follow is

acquire mutex
validate next stage
set StopChan and stage
release mutex

Otherwise somebody else could  do the same validation and take the same changes, resulting in two starts.  We want to show that exactly one of the possible goroutines starts apidump.

If we need to release the mutex (for example to make the CRI call) then when re-acquiring we have to re-validate any state we read while the mutex was first held.